File: tools/verify_fixture.py
#!/usr/bin/env python3
import sys
import json
from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.exceptions import InvalidSignature

V0_TEST_KEY_HEX = "37e9edc1ca6c423ec0955156b9bd318e7581ef4492b28a92235ee900d53174cc"


def load_json(path):
    with open(path, 'r') as f:
        return json.load(f)


def canonical_json(obj):
    """Deterministic canonical JSON for signing (RFC8785-like via sort_keys + compact)."""
    return json.dumps(obj, sort_keys=True, separators=(',', ':')).encode('utf-8')


def is_hex_32_bytes(value):
    if not isinstance(value, str):
        return False
    if len(value) != 64:
        return False
    try:
        bytes.fromhex(value)
        return True
    except ValueError:
        return False


def resolve_did_key(did):
    """Minimal V1 DID resolver stub.

    V1 only defines deterministic failure modes for now. Real did:key
    multicodec/multibase decoding is intentionally deferred.
    """
    if not isinstance(did, str) or not did:
        return None, "FAIL: did resolution failed"
    if did.startswith("did:key:"):
        return None, "FAIL: did resolution unsupported"
    return None, "FAIL: did resolution unsupported"


def resolve_public_key(receipt):
    """Resolve Ed25519 public key bytes.

    V1 order:
    1. proof.public_key
    2. proof.did
    3. FAIL: public key missing

    V0 compatibility:
    - V0 receipts may fall back to the fixed conformance key.
    """
    proof = receipt.get("proof", {})
    receipt_type = receipt.get("receipt_type")

    public_key_hex = proof.get("public_key")
    if public_key_hex:
        if public_key_hex == "V1_PUBLIC_KEY_HEX_PENDING":
            return None, "FAIL: public key missing"
        if not is_hex_32_bytes(public_key_hex):
            return None, "FAIL: public key invalid"
        return bytes.fromhex(public_key_hex), None

    did = proof.get("did")
    if did:
        if did == "did:key:V1_DID_KEY_PENDING":
            return None, "FAIL: did resolution unsupported"
        return resolve_did_key(did)

    if receipt_type == "AGENT_DELEGATION_RECEIPT_V0":
        return bytes.fromhex(V0_TEST_KEY_HEX), None

    return None, "FAIL: public key missing"


def verify_signature(receipt):
    """V0 + V1 Ed25519 verification."""
    try:
        proof = receipt.get("proof", {})
        sig_str = proof.get("signature", "")
        if not sig_str.startswith("ed25519:"):
            return False, "FAIL: signature mismatch"

        sig_hex = sig_str[8:]
        signature = bytes.fromhex(sig_hex)

        receipt_for_signing = {k: v for k, v in receipt.items()}
        receipt_for_signing["proof"] = {k: v for k, v in proof.items() if k != "signature"}
        message = canonical_json(receipt_for_signing)

        public_key_bytes, error = resolve_public_key(receipt)
        if error:
            return False, error

        public_key = ed25519.Ed25519PublicKey.from_public_bytes(public_key_bytes)
        public_key.verify(signature, message)
        return True, None
    except InvalidSignature:
        return False, "FAIL: signature mismatch"
    except ValueError:
        return False, "FAIL: public key invalid"
    except Exception:
        return False, "FAIL: signature mismatch"


def verify(receipt_path, binding_path, policy_path):
    try:
        receipt = load_json(receipt_path)
        binding = load_json(binding_path)
        policy = load_json(policy_path)

        receipt_type = receipt.get("receipt_type")
        receipt_version = receipt.get("receipt_version")

        if receipt_type not in ["AGENT_DELEGATION_RECEIPT_V0", "AGENT_DELEGATION_RECEIPT_V1"]:
            return "FAIL: schema invalid - wrong receipt_type"
        if receipt_type == "AGENT_DELEGATION_RECEIPT_V0" and receipt_version != "0.0.1":
            return "FAIL: schema invalid - wrong receipt_version"
        if receipt_type == "AGENT_DELEGATION_RECEIPT_V1" and receipt_version != "1.0.0":
            return "FAIL: schema invalid - wrong receipt_version"

        proof = receipt.get("proof", {})
        if "signature" not in proof:
            return "FAIL: schema invalid - missing signature in proof"

        required_binding = ["receipt_digest", "observed_files", "result_hash"]
        for field in required_binding:
            if field not in binding:
                return f"FAIL: schema invalid - missing {field} in binding"

        if not isinstance(policy.get("allowed_paths", []), list):
            return "FAIL: schema invalid - allowed_paths must be array"
        if not isinstance(policy.get("forbidden_paths", []), list):
            return "FAIL: schema invalid - forbidden_paths must be array"

        sig_valid, error = verify_signature(receipt)
        if not sig_valid:
            return error

        if binding.get("receipt_digest") != proof.get("digest"):
            return "FAIL: receipt digest mismatch"

        observed = set(binding.get("observed_files", []))
        allowed = set(policy.get("allowed_paths", []))
        forbidden = set(policy.get("forbidden_paths", []))

        if observed & forbidden:
            forbidden_file = next(iter(observed & forbidden))
            return f"FAIL: forbidden file touched: {forbidden_file}"

        if not observed.issubset(allowed):
            return "FAIL: scope violation - unauthorized file touched"

        return "PASS"

    except Exception as e:
        return f"FAIL: error - {str(e)}"


if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python verify_fixture.py <receipt.json> <binding.json> <policy.json>")
        sys.exit(1)

    result = verify(sys.argv[1], sys.argv[2], sys.argv[3])
    print(result)

File: tools/verify_fixture.js
#!/usr/bin/env node
import fs from 'node:fs';
import crypto from 'node:crypto';
import { fileURLToPath } from 'node:url';

function loadJson(path) {
  return JSON.parse(fs.readFileSync(path, 'utf8'));
}

function canonicalJson(obj) {
  // Deterministic canonical JSON - recursive key sort, compact (matches Python)
  if (obj === null || typeof obj !== 'object') {
    return JSON.stringify(obj);
  }
  if (Array.isArray(obj)) {
    return '[' + obj.map(canonicalJson).join(',') + ']';
  }
  const sortedKeys = Object.keys(obj).sort();
  const pairs = sortedKeys.map(key => {
    return JSON.stringify(key) + ':' + canonicalJson(obj[key]);
  });
  return '{' + pairs.join(',') + '}';
}

function verifySignature(receipt) {
  try {
    const proof = receipt.proof || {};
    const sigStr = proof.signature || '';
    if (!sigStr.startsWith('ed25519:')) return false;

    const sigHex = sigStr.slice(8);
    const signature = Buffer.from(sigHex, 'hex');

    const receiptForSigning = JSON.parse(JSON.stringify(receipt));
    receiptForSigning.proof = { ...proof };
    delete receiptForSigning.proof.signature;

    const canonicalString = canonicalJson(receiptForSigning);
    console.log('DEBUG_CANONICAL_HASH_NODE: ' + crypto.createHash('sha256').update(canonicalString, 'utf8').digest('hex'));
    const message = Buffer.from(canonicalString, 'utf8');

    const pubKeyHex = '37e9edc1ca6c423ec0955156b9bd318e7581ef4492b28a92235ee900d53174cc';
    const pubKeyBytes = Buffer.from(pubKeyHex, 'hex');

    // Ed25519 SPKI DER prefix for raw 32-byte public key
    const spkiHeader = Buffer.from('302a300506032b6570032100', 'hex');
    const derKey = Buffer.concat([spkiHeader, pubKeyBytes]);

    const publicKey = crypto.createPublicKey({
      key: derKey,
      format: 'der',
      type: 'spki'
    });

    return crypto.verify(null, message, publicKey, signature);
  } catch (e) {
    console.error('Signature verify error:', e.message);
    return false;
  }
}

function verify(receiptPath, bindingPath, policyPath) {
  try {
    const receipt = loadJson(receiptPath);
    const binding = loadJson(bindingPath);
    const policy = loadJson(policyPath);

    // === SCHEMA ENFORCEMENT ===
    if (receipt.receipt_type !== 'AGENT_DELEGATION_RECEIPT_V0') {
      return 'FAIL: schema invalid - wrong receipt_type';
    }
    if (receipt.receipt_version !== '0.0.1') {
      return 'FAIL: schema invalid - wrong receipt_version';
    }

    const proof = receipt.proof || {};
    if (!proof.signature) {
      return 'FAIL: schema invalid - missing signature in proof';
    }

    const requiredBinding = ['receipt_digest', 'observed_files', 'result_hash'];
    for (const field of requiredBinding) {
      if (!(field in binding)) {
        return `FAIL: schema invalid - missing ${field} in binding`;
      }
    }

    if (!Array.isArray(policy.allowed_paths)) {
      return 'FAIL: schema invalid - allowed_paths must be array';
    }
    if (!Array.isArray(policy.forbidden_paths)) {
      return 'FAIL: schema invalid - forbidden_paths must be array';
    }

    // === CRYPTO + REPLAY LOGIC ===
    if (!verifySignature(receipt)) {
      return 'FAIL: signature mismatch';
    }

    // Binding digest match
    if (binding.receipt_digest !== receipt.proof.digest) {
      return 'FAIL: receipt digest mismatch';
    }

    // Policy enforcement
    const observed = new Set(binding.observed_files || []);
    const allowed = new Set(policy.allowed_paths || []);
    const forbidden = new Set(policy.forbidden_paths || []);

    const forbiddenHit = [...observed].find(f => forbidden.has(f));
    if (forbiddenHit) {
      return `FAIL: forbidden file touched: ${forbiddenHit}`;
    }

    const unauthorized = [...observed].find(f => !allowed.has(f));
    if (unauthorized) {
      return 'FAIL: scope violation - unauthorized file touched';
    }

    return 'PASS';
  } catch (e) {
    return `FAIL: error - ${e.message}`;
  }
}

const isMain = process.argv[1] === fileURLToPath(import.meta.url);

if (isMain) {
  if (process.argv.length !== 5) {
    console.error('Usage: node verify_fixture.js <receipt.json> <binding.json> <policy.json>');
    process.exit(1);
  }

  const result = verify(process.argv[2], process.argv[3], process.argv[4]);
  console.log(result);
  process.exit(result === 'PASS' || result === 'FAIL: signature mismatch' ? 0 : 1);
}

export { verify, verifySignature, canonicalJson };
